internal/services/ingest.go

```diff
@@ -72,37 +80,64 @@ func NewIngestService(
 }
 
 func (m *ingestService) Run(ctx context.Context, startLedger uint32, endLedger uint32) error {
-	heartbeat := make(chan any)
-	go trackServiceHealth(heartbeat, m.appTracker)
+	rpcHeartbeat := make(chan entities.RPCGetHealthResult, 1)
+	ingestHeartbeat := make(chan any, 1)
+
+	// Start service health trackers
+	go trackRPCServiceHealth(ctx, rpcHeartbeat, m.appTracker, m.rpcService)
+	go trackIngestServiceHealth(ctx, ingestHeartbeat, m.appTracker)
+
 	if startLedger == 0 {
 		var err error
 		startLedger, err = m.models.Payments.GetLatestLedgerSynced(ctx, m.ledgerCursorName)
 		if err != nil {
 			return fmt.Errorf("erorr getting start ledger: %w", err)
 		}
 	}
+
 	ingestLedger := startLedger
-	for ; endLedger == 0 || ingestLedger <= endLedger; ingestLedger++ {
-		time.Sleep(10 * time.Second)
-		ledgerTransactions, err := m.GetLedgerTransactions(int64(ingestLedger))
-		if err != nil {
-			log.Error("getTransactions: %w", err)
-			continue
-		}
-		heartbeat <- true
-		err = m.ingestPayments(ctx, ledgerTransactions)
-		if err != nil {
-			return fmt.Errorf("error ingesting payments: %w", err)
-		}
+	for endLedger == 0 || ingestLedger <= endLedger {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancelled: %w", ctx.Err())
+		case resp := <-rpcHeartbeat:
+			switch {
+			// Case-1: wallet-backend is running behind rpc's oldest ledger. In this case, we start
+			// ingestion from rpc's oldest ledger.
+			case ingestLedger < resp.OldestLedger:
+				ingestLedger = resp.OldestLedger
+			// Case-2: rpc is running behind wallet-backend's latest synced ledger. We wait for rpc to
+			// catch back up to wallet-backend.
+			case ingestLedger > resp.LatestLedger:
+				log.Debugf("waiting for RPC to catchup to ledger %d (latest: %d)",
+					ingestLedger, resp.LatestLedger)
+				time.Sleep(5 * time.Second)
+				continue
+			}
+			log.Ctx(ctx).Infof("ingesting ledger: %d", ingestLedger)
 
-		err = m.processTSSTransactions(ctx, ledgerTransactions)
-		if err != nil {
-			return fmt.Errorf("error processing tss transactions: %w", err)
-		}
+			ledgerTransactions, err := m.GetLedgerTransactions(int64(ingestLedger))
+			if err != nil {
+				log.Error("getTransactions: %w", err)
+				continue
+			}
+			ingestHeartbeat <- true
+			err = m.ingestPayments(ctx, ledgerTransactions)
+			if err != nil {
+				return fmt.Errorf("error ingesting payments: %w", err)
+			}
```

Comment:
I'm assuming `ingestPayments` is a function that ingests all _inbound_ payments, since we have another function `processTSSTransactions` that I assume ingests all _outbound_ payments? Maybe we should update the function names to make that more clear?

Actually, I'm assuming both functions commit changes to the DB, which means that its possible to process _inbound_ payments for a ledger and fail to process the _outbound_ payments. If thats true, wouldn't this result in a bad state where we wouldn't be able to reingest the ledger, since retries would fail at the _inbound_ processing step?

Reply:
I would let @gouthamp-stellar answer here but yes you are correct that both these functions involve changes to db. So in order to avoid any of the issues you mentioned, we should use a single db transaction for both the functions.

@gouthamp-stellar any thoughts?

Reply:
ingestPayments ingests inbound and outbound payments to (and from) a stellar account. I don't think we should update the names of these functions because they do exactly what the name suggests. processTSSTransactions for example looks for transactions that have been submitted by TSS and only updates those particular transactions in our tables. 

And yes, they both do commit to the DB, but to a completely different set of tables - so I don't think they should all be bundled up into one atomic transaction. The way that the ingestor works is that if even one of these functions fails, the ingestor ingests the same ledger again, which is fine, because both these functions are idempotent

Reply:
So if I submit a payment transaction to the TSS for one of the accounts registered with the wallet backend, there will be two records for the same payment right? One provided by the TSS' transaction history, and one provided by the `/payments` endpoint.